test/integration.py

```diff
@@ -33,46 +34,46 @@ def site(request):
     container = f"mwclient-{tag}"
     # create the container, using upstream's official image. see
     # https://hub.docker.com/_/mediawiki
-    args = (_CONTAINER_RUNTIME, "run", "--name", container, "-p", f"{port}:80",
-            "-d", f"docker.io/library/mediawiki:{tag}")
+    args = [_CONTAINER_RUNTIME, "run", "--name", container, "-p", f"{port}:80",
```

Comment:
why change all of these? I habitually use tuples for cases where the thing doesn't need to be mutable, on the theory it's possibly slightly faster or more memory-efficient or something. does it cause issues for typing? typeshed doesn't seem to expect these to be lists specifically, just sequences...

Reply:
The issue is not with the signature of `subprocess.run`, but rather with the redefinition of the `args` variable using tuples of different lengths:

```
test/integration.py:43: error: Incompatible types in assignment (expression has type tuple[str, str, ... <17 more items>], variable has type tuple[str, str, str, str, str, str, str, str])  [assignment]
test/integration.py:49: error: Incompatible types in assignment (expression has type tuple[str, str, ... <9 more items>], variable has type tuple[str, str, str, str, str, str, str, str])  [assignment]
test/integration.py:54: error: Incompatible types in assignment (expression has type tuple[str, str, ... <12 more items>], variable has type tuple[str, str, str, str, str, str, str, str])  [assignment]
...
```

The arity of a tuple is part of its type (e.g., `tuple[str, str]` is a different type from `tuple[str, str, str]`), so essentially, the type of the `args` variable keeps changing. From mypy’s perspective, this is equivalent to the following, which would clearly fail type-checking:

```python
foo = 1
foo = "bar"
foo = True
```

To work around this would  either
1. have to set `allow_redefinition = true` in the `tools.mypy` section of the `pyproject.toml` (that would also allow the stuff like my previous example though)
2. use different variables (i.e. `args1`, `args2`, …)
3. use lists instead as I did here. The length of a list is not part of its type, so everything works as it should

You're right that tuples would (probably) be slightly more efficient than lists here, but in the context here I think it doesn't really matter.

Reply:
I see, that makes sense. Thanks!